plutus-use-cases/src/Language/Marlowe/Compiler.lhs

```diff
@@ -0,0 +1,841 @@
+                    (i, (_, NotRedeemed val _)) : ls | i `eqIdentCC` id1 && change == scriptValue - val ->
+                        findAndRemove ls resultCommits (True, state)
+                    e : ls -> findAndRemove ls (e : resultCommits) result
+                    [] -> let
+                        (isValid, State _ choices) = result
+                        in (isValid, State (reverse resultCommits) choices)
+
+                (ok, updatedState) = findAndRemove commits [] (False, state)
+                isValid = ok
+                    && validatorHash `eqValidator` thisScriptHash
+                in if isValid
+                then (updatedState, con, True)
+                else (state, contract, False)
+
+            (_, Redeem identCC) -> let
+                    PendingTx [PendingTxIn _ _ (Plutus.Value scriptValue)]
+                        (PendingTxOut (Plutus.Value change) (Just (validatorHash, _)) DataTxOut : _)
+                        _ _ _ _ thisScriptHash = pendingTx
+
+                    findAndRemoveExpired ::
+                        [(IdentCC, CCStatus)]
+                        -> [(IdentCC, CCStatus)]
+                        -> (Bool, State)
+                        -> (Bool, State)
+                    findAndRemoveExpired ls resultCommits result = case ls of
+                        (i, (_, NotRedeemed val expire)) : ls |
+                            i `eqIdentCC` identCC && change == scriptValue - val && currentBlockNumber > expire ->
+                                findAndRemoveExpired ls resultCommits (True, state)
+                        e : ls -> findAndRemoveExpired ls (e : resultCommits) result
+                        [] -> let
+                            (isValid, State _ choices) = result
+                            in (isValid, State (reverse resultCommits) choices)
+
+                    (ok, updatedState) = findAndRemoveExpired commits [] (False, state)
+                    isValid = ok
+                        && validatorHash `eqValidator` thisScriptHash
+                    in if isValid
+                    then (updatedState, contract, True)
+                    else (state, contract, False)
+
+            (Null, SpendDeposit) | null commits -> (state, Null, True)
+
+            _ -> (state, Null, False)
+
+        contractIsValid = validateContract [] [] marloweContract
+
+        -- record Choices from Input into State
+        stateWithChoices = let
+            State commits choices = marloweState
+            in State commits (mergeChoices inputChoices choices)
+
+        (_::State, _::Contract, allowTransaction) = eval inputCommand stateWithChoices marloweContract
+        -- if a contract is not valid we allow a contract creator to spend its initial deposit
+        -- otherwise, if the contract IS valid we check the contract allows this transaction
+        in if not contractIsValid || allowTransaction then ()
+        else Builtins.error ()
+        ||])
```

Comment:
This might be the most complex validator script ever written in Plutus!

Reply:
I think it is. Thank you for your help!